File: Dynamic Programming/Best Time to Buy and Sell Stock.java
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/description/
We can buy on min price day and sell everyday to check maximum profit.
class Solution {
    public int maxProfit(int[] prices) {
        int max = 0;
        int min = prices[0];
        for(int i=1; i<prices.length; i++){
            min = Math.min(min,prices[i]);
            max = Math.max(max, prices[i]-min);
        }
        return max;
    }
}

File: Dynamic Programming/Best Time to Buy and Sell Stock.ts
//https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

function maxProfit(prices: number[]): number {
    if(prices.length <= 1) return 0; 
    let buy: number = prices[0];
    let sell: number = -1e9;
    let profit: number = 0;
    for(let i = 1; i < prices.length; i++){
        profit = Math.max(prices[i] - buy, profit);
        if(prices[i] < buy)
            buy = prices[i];
    }
    return profit;
};